### DISCUSSION TITLE
Sandboxed workers: best way to setup DI container with long running TCP connections in services (like DB connection, Redis connection, etc.)

### DISCUSSION THREAD

**Discussion by tolgap** [taskforcesh/bullmq#942] [Q&A] [answered]
I have a custom application that uses BullMQ workers in sandboxed mode. I have 2 queues in use. These 2 queues are handled asynchronously in separate processes (sandboxed).

I'm having trouble figuring out how to handle the DB connection. To be precise, it's more about when to close the DB. Right now, I'm using the following logic in my BullMQ processor entry point:

```ts
export default async function Processor(job: Job<BaseJobData>) {
  // create inversify container with all DB repo's and service classes that inject each other.
  const container = await initContainer();
  const connection = container.get<MikroORM<IDatabaseDriver<Connection>>>(
    TYPES.DATABASE_CONNECTION
  );

  // create the Databse Context so we don't use the global DB context.
  // Within this context, we execute the BullMQ Job with `performJob`.
  await RequestContext.createAsync(connection.em, async function () {
    await performJob(job);
  });

  // I would like to call `connection.close(true)` here so the worker process can exit gracefully
  await connection.close(true);
}
```

But when I call `connection.close()` at the end of my processor, it seems that BullMQ can "reuse" this process for another Job in the Queue that is waiting. And then, I have a DB connection which has been closed, and cannot be reused again.

What logic would be necessary to not have dangling worker processes that keep a connection alive to the DB in the context of sandboxed BullMQ workers? Anyone got tips?

**Comment by manast** [comment, marked as answer]
I think the best is to create your connection outside of the processor so that it gets reused every time the process function is called, and then close the connection on the SIGINT event listener for the process:
```ts
process.on('SIGINT', await function () {
    await connection.close();
    process.exit();
  });
```

https://nodejs.org/api/process.html#signal-events

**Reply by tolgap** [reply]
Well that actually works really well. Thank you!
